pp_ctl.c - make sure we dont leak PL_restartop from failed compile

demerphq opened the pull request:
Fix GH Issue #20396, try_yyparse() breaks Attribute::Handlers.

Reduced test case is:

```
    perl -e'CHECK { eval "]" }'
```

which should not assert or segfault.

In c304acb49 we made it so that when doeval_compile() is executed and it calls yyparse() inside of an eval any exceptions that occur during the parse process are trapped by try_yyparse() so that exection would return to doeval_compile(). This was done so that post eval compilation cleanup logic could be handled similarly regardless of whether Perl_croak() was called or not. However the logic to setup PL_restartop was not adjusted accordingly.

The opcode that calls doeval_compile() setups an eval context data before it calls doeval_compile().  This data includes the "retop" which is used to return control to after the eval should it die and is set to the be the evaling opcodes op_next. When Perl_die_unwind() is called it sets PL_restartop to be the "retop" of the of the current eval frame, and then does a longjmp, on the assumption it will end up inside of a "run loop enabled jump enviornment", where it restarts the run loop based on the value of PL_restartop, zeroing it aftewards.

After c304acb49 however, a die inside of try_yyparse the die_unwind returns control back to the try_yyparse, which ignores PL_restartop, and leaves it set. Code then goes through the "compilation failed" branch and execution returns to PL_restartop /anyway/, as PL_op hasn't changed and pp_entereval returns control to PL_op->op_next, which is what we pushed into the eval context anyway for the PL_restartop.

The end result of this however is that PL_restartop remains set when we enter perl_run() for the first time. perl_run() is a "run loop enabled jump enviornment" which uses run_body() to do its business, such that when PL_restartop is NULL it executes the just compiled body of the program, and when PL_restartop is not null it assumes it must be in the eval handler from an eval from the main body and it should recontinue. The leaked PL_restartop is thus executed instead of the main program body and things go horribly wrong.

This patch changes it so that when try_yyparse traps an exception we restore PL_restartop back to its old value. Same for its partner PL_restartjmpenv. This is fine as they have been set to the values from the beginning of the eval frame which we are part of, which is now over.

demerphq committed:
pp_ctl.c - in try_yyparse do not leak PL_restartop from compile that dies

Fix GH Issue #20396, try_yyparse() breaks Attribute::Handlers.

Reduced test case is:

    perl -e'CHECK { eval "]" }'

which should not assert or segfault.

In c304acb49 we made it so that when doeval_compile() is executed and it
calls yyparse() inside of an eval any exceptions that occur during the
parse process are trapped by try_yyparse() so that exection would return
to doeval_compile(). This was done so that post eval compilation cleanup
logic could be handled similarly regardless of whether Perl_croak() was
called or not. However the logic to setup PL_restartop was not adjusted
accordingly.

The opcode that calls doeval_compile() setups an eval context data
before it calls doeval_compile(). This data includes the "retop" which
is used to return control to after the eval should it die and is set to
the be the evaling opcodes op_next. When Perl_die_unwind() is called it
sets PL_restartop to be the "retop" of the of the current eval frame,
and then does a longjmp, on the assumption it will end up inside of a
"run loop enabled jump enviornment", where it restarts the run loop
based on the value of PL_restartop, zeroing it aftewards.

After c304acb49 however, a die inside of try_yyparse the die_unwind
returns control back to the try_yyparse, which ignores PL_restartop, and
leaves it set. Code then goes through the "compilation failed" branch
and execution returns to PL_restartop /anyway/, as PL_op hasn't changed
and pp_entereval returns control to PL_op->op_next, which is what we
pushed into the eval context anyway for the PL_restartop.

The end result of this however is that PL_restartop remains set when we
enter perl_run() for the first time. perl_run() is a "run loop enabled
jump enviornment" which uses run_body() to do its business, such that
when PL_restartop is NULL it executes the just compiled body of the
program, and when PL_restartop is not null it assumes it must be in the
eval handler from an eval from the main body and it should recontinue.
The leaked PL_restartop is thus executed instead of the main program
body and things go horribly wrong.

This patch changes it so that when try_yyparse traps an exception we
restore PL_restartop back to its old value. Same for its partner
PL_restartjmpenv. This is fine as they have been set to the values from
the beginning of the eval frame which we are part of, which is now over.

bram-perl commented on dist/Attribute-Handlers/lib/Attribute/Handlers.pm:
```diff
@@ -249,12 +249,16 @@ sub _apply_handler_AH_ {
-		my $evaled = eval("package $pkg; no warnings; no strict;
-				   local \$SIG{__WARN__}=sub{die}; [$data]");
-		$data = $evaled unless $@;
+                my $code= "package $pkg; my \$ref= [$data]; \$data= \$ref; 1";
+                local $SIG{__WARN__} = sub{ die };
+                no strict;
+                no warnings;
+                eval($code);
```
I would probably move the Attribute-Handlers change to it's own commit. (These make the code more readable but shouldn't be required for the actual bug fix)

(and/or maybe add something like: `# print STDERR "Evaling code: $code\n";` since it could be useful when debugging. (only suggesting it because there already are commented `print STDERR` statements in the code, a debug of verbose flag might be more useful but might be overkill))

demerphq replied:
Sure. NO problem. Ironically i deleted the print STDERR stuff I had in there before pushing. Ill put it back. :-)

demerphq replied:
Done in the latest push.

bram-perl commented on perl.c:
```diff
@@ -2770,6 +2775,10 @@ S_run_body(pTHX_ I32 oldscope)
     PERL_SET_PHASE(PERL_PHASE_RUN);
 
     if (PL_restartop) {
+#ifdef DEBUGGING
+        if (!DEBUG_q_TEST)
+          PERL_DEBUG(PerlIO_printf(Perl_debug_log, "\nRESTARTOP...\n\n"));
+#endif
```
Did you intend to include this?
(The debug message itself doesn't tell very much; if it was intended: can the message be improved?)

demerphq replied:
Yeah, I did, but i can tweak it. That shouldnt happen, and i wanted to make it obvious when it does.

bram-perl commented on t/op/eval.t:
```diff
@@ -723,10 +723,10 @@ pass("eval in freed package does not crash");
         my $sort_line= 'my @x= sort { ' . $line . ' } 1,2;';
-        fresh_perl_is($line . ' print "ok";', "A-ok", {}, "No segfault: $line");
-
+        fresh_perl_is($line . ' print "ok";', "A-ok", {}, "No segfault inside sort: $line");
```
Hmm.. this doesn't make sense (but neither did the original code)
I think it should be: `fresh_perl_is($sort_line . ' print.... )`
Right now it appears to be calling `fresh_perl_is($line...`) twice

demerphq replied:
Doh. Good catch.

demerphq replied:
Fixed in the lastest push.

bram-perl commented on t/op/eval.t:
```diff
@@ -723,10 +723,10 @@ pass("eval in freed package does not crash");
         'eval "UNITCHECK { eval q(UNITCHECK { die; }); print q(A-) }";',
         'eval "UNITCHECK { eval q(BEGIN     { die; }); print q(A-) }";',
         'eval "BEGIN     { eval q(UNITCHECK { die; }); print q(A-) }";',
+        'CHECK { eval "]" } print q"A-";'
```
Might be good to also test: `INIT { eval "]" } ...` since that  was observed (by @haarg) to segfault as well.

demerphq replied:
Yeah, willdo.

bram-perl commented on pp_ctl.c:
```diff
@@ -3724,6 +3724,12 @@ S_doeval_compile(pTHX_ U8 gimme, CV* outside, U32 seq, HV *hh)
     /* compile the code */
+
+    /* if we die during compilation PL_restartop and PL_restartjmpenv
+     * will be set. We need to restore their values if that happens. */
```
Consider changing it to something like:
```
    /* if we die during compilation PL_restartop and PL_restartjmpenv
     * will be set (by Perl_die_unwind). We need to restore their values if that happens. */
```

demerphq replied:
Sure.

demerphq replied:
Fixed in latest push.

bram-perl commented on pp_ctl.c:
```diff
@@ -3733,10 +3739,19 @@ S_doeval_compile(pTHX_ U8 gimme, CV* outside, U32 seq, HV *hh)
-         * various vars restored */
-        if (yystatus != 3) {
+        if (yystatus == 3) {
+            /* try_yyparse() died and we trapped the error. We need to restore
+             * the old PL_restartjmpenv and PL_restartop values. */
+            assert(!in_require); /* not in a require */
+            assert(PL_restartop == PL_op->op_next); /* we expect these to match */
+            PL_restartjmpenv = restartjmpenv;
+            PL_restartop = restartop;
```
Hmm.. right now I'm not yet convinced if this is the proper place to fix it.

What comes to mind:
- why is the fix not in `S_try_yyparse`?
- if `S_tty_yyparse` can cause the breakage then won't the same breakage be possible via `S_try_run_unitcheck`? (I haven't tested it yet)

demerphq replied:
I debated putting it in try_yyparse, but decided to leave it as a very clean light wrapper around yyparse.  I can do it the other way. Ill check S_try_rununitcheck when I get a chance.

demerphq replied:
I moved it to try_yyparse(), and added equivalent logic to S_try_rununitcheck(). I couldn't find a test case for the UNITCHECK scenario but if I added an assert about PL_restartop and I could see it "leaking" into S_try_rununitcheck where I believe it shouldn't so I added it to that as well.

demerphq committed:
Attribute-Handlers: rework eval code

This makes the logic a bit simpler, and makes it easier to debug
as well. Reducing the amount of code that needs to be inside
the eval makes it easier to debug what is going on, especially
from an internals point of view (eg with -Dl enabled).

demerphq committed:
perl.c - add debug output to show when the run_body() continues after eval fail

This can be helpful debugging issues related to phaser blocks and eval.

demerphq committed:
scope.c - sanity check a var before we use it

demerphq committed:
pp_ctl.c - in try_run_unitcheck() guard against leaking PL_restartop

If we die while executing a UNITCHECK inside of an eval we shouldn't
leave PL_restartop set, as we will execute PL_op->op_next anyway. See
the previous commit for more details in the context yy_parse().

Thanks to Bram for coming up with a test case that demonstrated the
problem.

demerphq commented:
@bram-perl fixed and pushed. Thanks for the test case.

demerphq commented:
Unless someone objects I will merge this bug fix tomorrow without approval, or sooner if someone acks the patch.

tonycoz commented on dist/Attribute-Handlers/lib/Attribute/Handlers.pm:
```diff
@@ -241,20 +241,33 @@ sub _apply_handler_AH_ {
+                local $SIG{__WARN__} = sub{ die };
+                no strict;
+                no warnings;
+                # Note we do not need
```
What don't we need?

demerphq replied:
Oh. Drat. Will fix. Thanks!

demerphq replied:
Fixed. Thanks! Replaced with:
```
+                # Note in production we do not need to use the return value from
+                # the eval or even consult $@ after the eval - if the evaled code
+                # compiles and runs successfully then it will update $data with
+                # the compiled form, if it fails then $data stays unchanged. The
+                # return value and $@ are only used for debugging purposes.
+                # IOW we could just replace the following with eval($code);
```

tonycoz commented:
Except for the incomplete comment it looks good to me.
